### PR TITLE
[#1498] Implement Camera Bounds Strategy

### LINKED ISSUE #1498: Allow for camera constraints
### Context
Right now, a camera can move anywhere and everywhere. In the context of showing a map, this will mean that when bound to an Actor, the camera moves beyond the limits of the map and shows sudden emptiness.

### Proposal
I propose to add an extra constraints property to the CameraStrategy class, that, when set, will stop the camera once it goes beyond the defined constraints.

Example:
```ts
this.cameraStrategy.constraints = new Vector(mapWidth, mapHeight);
```
CameraStrategy.action() is then expected to follow the constraints and return the appropriate value.

### PR DESCRIPTION
**Pull request by DaVince**
===:clipboard: PR Checklist :clipboard:===

- [x] :pushpin: issue exists in github for these changes
- [x] :microscope: existing tests still pass
- [x] :see_no_evil: code conforms to the [style guide](https://github.com/excaliburjs/Excalibur/blob/master/STYLEGUIDE.md)
- [x] :triangular_ruler: new tests written and passing / old tests updated with new scenario(s)
- [x] :page_facing_up: changelog entry added (or not needed)

Closes #1498

## Changes:

- Adds a new camera strategy named LimitCameraBoundsStrategy.
- Adds the strategy to the StrategyContainer.

### CODE REVIEW

**Comment by DaVince** [comment]
I'd like some feedback on this before I add tests.

Also, is that failed PR Labeler check going to be an issue? It only exists because I forgot to sync my fork before working on this...

**Comment by DaVince** [comment]
An example implementation is available at https://github.com/DaVince/excalibur-tiled-examples/blob/master/src/index.ts#L24

File: src/engine/Camera.ts [hunk 1]
```diff
@@ -170,6 +178,35 @@ export class RadiusAroundActorStrategy implements CameraStrategy<Actor> {
   };
 }
 
+/**
+ *  Prevent a camera from going beyond the given camera dimensions.
+ */
+export class LimitCameraBoundsStrategy implements CameraStrategy<BoundingBox> {
```
Comment (eonarheim) [inline src/engine/Camera.ts:184]: @DaVince this is great, exactly what I was thinking for implementation

Comment (DaVince) [reply src/engine/Camera.ts:184]: Cool, I'll start writing some tests then!

**Comment by DaVince** [comment]
Added tests, changelog, more documentation.

I'd like to note that this strategy does not transform any of the other strategies or camera actions. So say you try to `move()` the camera beyond the boundaries with an easing function, the strategy won't change the `move()` target to your bounding box limits. This can cause a jarring effect (sudden stop) but I figure that it's better dealt with within a game dev's own code. Maybe I can implement it later as an optional feature.

**Comment by DaVince** [comment]
I need to add some more tests to this that checks if it handles non-zero x and y start positions properly.

**Comment by DaVince** [comment]
Squashed a bug relating to my previous post.

I also observed crazy jittery behavior when the bounding box is smaller than the screen size. The only way I can fix that is by growing the bounding box on initialization. Is there any way to do that without passing an instance of `Engine` to the constructor? I don't want to run this logic on every `action()`.

**Comment by DaVince** [comment]
If that last thing is a non-issue, this PR is good to go. :)

**Comment by eonarheim** [comment]
> Squashed a bug relating to my previous post.
> 
> I also observed crazy jittery behavior when the bounding box is smaller than the screen size. The only way I can fix that is by growing the bounding box on initialization. Is there any way to do that without passing an instance of `Engine` to the constructor? I don't want to run this logic on every `action()`. (Or maybe I _should,_ considering a dev might change the size of the target bounding box?)

I think it's a non-issue, we could possibly log a warning about the bound box size and/or some documentation in the strategy about the expected bounds.

**Comment by DaVince** [comment]
I've added documentation and a warning.

The warning is implemented a little clumsily inside of the `action` method since I don't have access to Engine properties in the constructor.

**Comment by eonarheim** [comment]
@DaVince thanks again for your contributions!
